docs/constrained-generators/manual.md

```diff
@@ -330,119 +498,979 @@ So would any type whose HasSpec instance was derived via the GHC.Generics instan
+width_ d = sel @1 d   
+```
+
+This requires the `DataKinds` directive, and importing `GHC.Generics` and `GHC.TypeLits` to work.
+
+```
+{-# LANGUAGE DataKinds #-}
+import GHC.Generics
+import GHC.TypeLits
+```
+
+When we type-apply the library function `sel` to a type-level `Natural` number like this `(sel @1)` it
+selects the `ith` selector function. The selectors are numbered from `0` to `n-1` . Selection 
+can always be expressed using `match` like this:
+
+```
+ex20a :: Specification Dimensions
+ex20a = constrained $ \ d ->
+        match d $ \ l w d -> [ l >. lit 10, w ==. lit 5, d <. lit 20]
+```
+
+which can be reexpressed using `sel` as this.
+
+```
+ex20b :: Specification Dimensions
+ex20b = constrained $ \ d ->
+        [sel @0 d >. lit 10
+        ,sel @1 d ==. lit 5
+        ,sel @2 d <. lit 20]     
+```
+
+When we wish to constrain just a subset of the subcomponents, selectors make it possible 
+to write more concise `Specification`s.
+
+```
+ex21 :: Specification Dimensions
+ex21 = constrained $ \ d -> width_ d ==. lit 1
+```
+
+
+## Naming introduced lambda bound Term variables
+1.  [var|name|]
+
+When we use a library function that introduces new Term variable using a Haskell lambda expression, the system
+gives the Haskell variable a unique Term level name such as `v0` or `v1`  or `v2` etc. When the specification is over constrained, the system attempts to explain what is wrong, but when doing this, it uses the internal Term level names that were assigned to the Haskell variables. This means the error messages are often hard to
+understand. For example consider the over constrained specification. Over constrained because
+`left` cannot be simultaneously equal to `right` and `right + lit 1`
+
+```
+ex22a :: Specification (Int,Int)
+ex22a = constrained $ \ pair -> 
+       match pair $ \ left right ->
+	      [ left ==. right, left ==. right + lit 1]
+```
+
+When attempt to solve this specitication for `pair` we get the following error message
+
+```
+debugSpec ex22a
+StepPlan for variable: v_1 fails to produce Specification, probably overconstrained. 
+Original spec ErrorSpec
+  Intersecting: 
+    MemberSpec [20]
+    MemberSpec [19]
+  Empty intersection
+```
+
+Note the error message is in terms of the internal Term name `v1`. Which is not very useful.
+
+To make error messages clearer we can name Haskell lambda bound variables using `[var|left|]` instead of just `left`.  In order to do this we must have the following directives in our file.
+
+```
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE ViewPatterns #-}
+```
+
+Here is the same example using this naming schema.
+
+```
+ex22b :: Specification (Int,Int)
+ex22b = constrained $ \ [var|pair|] -> 
+         match pair $ \ [var|left|] [var|right|] -> [ left ==. right, left ==. right + lit 1]
+```
+
+We now get a better error message
+
+```
+ghci> debugSpec ex22b
+StepPlan for variable: left_1 fails to produce Specification, probably overconstrained.PS = 
```

Comment:
```suggestion
StepPlan for variable: left_1 fails to produce Specification, probably overconstrained.
```
The previous error message didn't have this, so I assume this was an accident.

Reply:
yes, I thin that came from cut and paste, to move sections around. Thanks for pointing it out.